### PR TITLE
Add health check API

### LINKED ISSUE #721: Health endpoint
should be based on main

### PR DESCRIPTION
**Pull request by yaziciahmet**
# Description
Health check API is added to both sequencer and full node. `ProxyGetRequestLayer` from jsonrpsee crate is used to proxy `GET /health` requests to the `health_check` rpc method, which just returns `Ok` value.

## Linked Issues
- Fixes #721 

## Testing
I ran a sequencer and a fullnode, and sent a GET request to `/health`.

### CODE REVIEW

File: da-db/mock_da.db
[the remark this answers is not in the document]
Comment (yaziciahmet) [reply da-db/mock_da.db:1]: mb :) deleted https://github.com/chainwayxyz/citrea/pull/834/commits/40a8361e3e6961e4fc674de558e31e56827113a8

File: crates/sequencer/src/sequencer.rs [hunk 1]
```diff
@@ -918,14 +922,50 @@ where
...
         let rpc_context = self.create_rpc_context().await;
-        let rpc = create_rpc_module(rpc_context)?;
-        rpc_methods.merge(rpc)?;
-        Ok(rpc_methods)
+        let mut rpc = create_rpc_module(rpc_context)?;
+
+        // TODO: create register_healthcheck_rpc helper function
+        let ledger_db = self.ledger_db.clone();
+        rpc.register_method("health_check", move |_, _| {
+            let next_soft_batch_num = ledger_db.get_next_items_numbers().soft_batch_number;
```
Comment (eyusufatik) [inline crates/sequencer/src/sequencer.rs:935]: using this method might be problematic. can't you use `ledger.getHeadSoftBatch()`?

File: crates/sequencer/src/sequencer.rs [hunk 2]
```diff
@@ -918,14 +922,50 @@ where
...
+                .map_err(|err| {
+                    ErrorObjectOwned::owned(
+                        INTERNAL_ERROR_CODE,
+                        INTERNAL_ERROR_MSG,
+                        Some(format!("Failed to get soft batch range: {}", err)),
+                    )
+                })?;
+
+            let block_time_s = soft_batches[1].timestamp - soft_batches[0].timestamp;
+            thread::sleep(Duration::from_millis(block_time_s * 1500));
```
Comment (eyusufatik) [inline crates/sequencer/src/sequencer.rs:953]: do async sleep with tokio, you can do `rpc.register_async_method()`

**Review by eyusufatik — CHANGES_REQUESTED** [review]
healthcheck is fine, left small nits

**Comment by eyusufatik** [comment]
also we should have a test like this:

in bin/citrea/tests/e2e

create a new test

set up a sequencer and full node, set the sequencer to test_mode = false so that it produces blocks on its own.

send health check request to the full node and check that it's 200.

abort the sequencer task and send health check request to the full node and check that it's 400/500 whatever

File: crates/sequencer/src/sequencer.rs [hunk 3]
```diff
@@ -931,36 +930,57 @@ where
...
-
-            let soft_batches = ledger_db
-                .get_soft_batch_range(
-                    &(BatchNumber(next_soft_batch_num - 2)..BatchNumber(next_soft_batch_num)),
-                )
-                .map_err(|err| {
+        rpc.register_async_method("health_check", move |_, _| {
+            let ledger_db = ledger_db.clone();
+            async move {
+                let head_batch = ledger_db.get_head_soft_batch().map_err(|err| {
```
Comment (eyusufatik) [inline crates/sequencer/src/sequencer.rs:936]: why not make the whole closure async move?

Comment (yaziciahmet) [reply crates/sequencer/src/sequencer.rs:936]: It caused all move and lifetime issues. I tried Arc and everything but only cloning `ledger_db` before every invocation of the async part of the code solved the issue.

Comment (eyusufatik) [reply crates/sequencer/src/sequencer.rs:936]: @kpp @rakanalh any better way to do this?

if not, @yaziciahmet make sure clone on ledger db is a simple operation

Comment (yaziciahmet) [reply crates/sequencer/src/sequencer.rs:936]: we can also use the RpcContext to share the ledger_db. The reason I didn't go that way initially is, full node is not using RpcContext but sequencer is using it, hence it was easier to clone the ledger_db and use it. I will try to make RpcContext shared in common crate and use it in both places.

Comment (eyusufatik) [reply crates/sequencer/src/sequencer.rs:936]: that makes more sense, however, the sequencer has mempool in its context and the prover does not. what about that?

Comment (yaziciahmet) [reply crates/sequencer/src/sequencer.rs:936]: Might be a bit dirty, but we can have a trait in common called `LedgerDBAccessor`, and the actual context implementations can be different for Sequencer and FullNode. And these contexts would implement this trait simply returning the LedgerDB.

Comment (eyusufatik) [reply crates/sequencer/src/sequencer.rs:936]: maybe.

your call :)

Comment (yaziciahmet) [reply crates/sequencer/src/sequencer.rs]: I believe it is unnecessary to deal with contexts right now as cloning LedgerDB seems to be really cheap due to its' `db` and `next_item_numbers` fields being encapsulated with an Arc, and as far as I researched, it is also a cheap operation to clone a broadcast channel. Hence I will keep the implementation as it is.

File: bin/citrea/tests/e2e/mod.rs [hunk 4]
```diff
@@ -3164,3 +3164,78 @@ async fn test_full_node_sync_status() {
...
+    let full_node_url = format!("http://localhost:{}/health", full_node_addr.port());
+    let resp = reqwest::get(&full_node_url).await.unwrap();
+    assert_eq!(resp.status().as_u16(), 200);
+
+    wait_for_l2_block(&full_node_test_client, 4, None).await;
+
+    seq_task.abort();
+
+    let resp = reqwest::get(&full_node_url).await.unwrap();
+    assert_eq!(resp.status().as_u16(), 500);
```
Comment (eyusufatik) [inline bin/citrea/tests/e2e/mod.rs:3238]: instead of using reqwest directly here, can you put health request to test_client.rs ?

**Review by eyusufatik — CHANGES_REQUESTED** [review]

File: bin/citrea/tests/e2e/mod.rs (hunk 4 above)
Comment (yaziciahmet) [reply bin/citrea/tests/e2e/mod.rs]: Done https://github.com/chainwayxyz/citrea/pull/834/commits/fdb5e83a496122a74aead8061c2489ab7270fbd0

**Review by eyusufatik — APPROVED** [review]

**Pull request merged**